src/connection.ts

```diff
@@ -4,70 +4,235 @@ import { AbortOptions } from '@libp2p/interfaces';
     this.peerConnection = init.pc;
-    this.ufrag = init.credential_string;
+    this.remotePeer = init.remotePeer;
+    this.localPeer = init.localPeer;
+    this.localAddr = init.localAddr;
+    this.components = init.components;
     this.stat = {
-      direction: 'outbound',
-      timeline: { open: 0 },
-      status: 'CLOSED',
+      direction: init.direction,
+      status: 'OPEN',
+      timeline: {
+        open: new Date().getTime(),
+      },
+    };
+    this.handleIncomingStreams();
+  }
+
+  private handleIncomingStreams() {
+    let metrics = this.components.getMetrics();
+    this.peerConnection.ondatachannel = async ({ channel }) => {
+      const logPrefix = `[stream:${channel.label}][inbound]`;
+      log.trace(`incoming stream - ${channel.label}`);
+      let [openPromise, abortPromise] = [defer(), defer()];
+      let controller = new TimeoutController(OPEN_STREAM_TIMEOUT);
+      controller.signal.onabort = () => abortPromise.resolve();
+      channel.onopen = () => openPromise.resolve();
+
+      await Promise.race([openPromise.promise, abortPromise.promise]);
+      if (controller.signal.aborted) {
+        // TODO: Better errors
+        throw Error(controller.signal.reason);
+      }
+
+      let rawStream = new WebRTCStream({
+        channel,
+        stat: {
+          direction: 'inbound',
+          timeline: {
+            open: new Date().getTime(),
+          },
+        },
+      });
+      let registrar = this.components.getRegistrar();
+      let protocols = registrar.getProtocols();
+
+      log.trace(`${logPrefix} supported protocols - ${protocols}`);
+
+      let { stream, protocol } = await mshandle(rawStream, protocols, { signal: controller.signal });
+      if (metrics) {
+        metrics.trackStream({ stream, protocol, remotePeer: this.remotePeer });
+      }
+
+      log.trace(`${logPrefix} handled protocol - ${protocol}`);
+
+      rawStream.stat.protocol = protocol;
+      let result = this.wrapMsStream(rawStream, stream);
+
+      this.addStream(result);
+
+      // handle stream
+      let { handler } = registrar.getHandler(protocol);
+      handler({ connection: this, stream: result });
     };
-    this.remotePeer = init.remotePeerId;
-    // for muxing incoming stream
-    // this._peerConnection.ondatachannel = ({ channel }) => {
-    // 	let stream = DataChannelStream(channel)
-    // 	this.addStream(stream)
-    // }
   }
 
-  async newStream(multicodecs: string | string[], options?: AbortOptions): Promise<ic.Stream> {
-    // let label = uuid.v4()
-    // let dc = this._peerConnection.createDataChannel(label, {})
-    // await datachannel opening
-    // return DataChannelStream(dc)
-    log('TODO', this.ufrag);
-    this.peerConnection.createDataChannel(genUuid());
-    throw new Error('not implemented');
+  private wrapMsStream(rawStream: WebRTCStream, stream: Duplex<Uint8ArrayList, Uint8ArrayList | Uint8Array, Promise<void>>): ic.Stream {
+    return {
```

Comment:
Is this wrapping just to be able to set the ID? Because if so... we could add that as an optional thing to StreamInit